tests/features/S2N_LIBCRYPTO_SUPPORTS_ENGINE.c

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/*
+ * This feature probe checks if the linked libcrypto has ENGINE support [1].
+ *
+ * Normally, we would always expect the `openssl/engine.h` header to be
+ * available. However, some platforms (CentOS 10,  Fedora 41, and RHEL 10 are
+ * manually removing the `openssl/engine.h` header (for more detail look at
+ * issues [2] [3]).
+ *
+ * [1] https://docs.openssl.org/1.0.2/man3/engine/
+ * [2] https://github.com/aws/s2n-tls/pull/4705
+ * [3] https://github.com/aws/s2n-tls/pull/4873
+ */
+
+#include <openssl/engine.h>
+
+int s2n_openssl_compat_rand(unsigned char *buf, int num)
+{
+    return 1;
+}
+
+int main()
+{
+#if defined(OPENSSL_NO_ENGINE)
+    #error "engine is not supported"
+#endif
```

Comment:
Moving OPENSSL_NO_ENGINE to a runtime check doesn't mean that we are assuming that the engine apis will be available if OPENSSL_NO_ENGINE is defined. The feature probe checks for the existence of the engine apis and gates them if necessary. The feature probe doesn't need OPENSSL_NO_ENGINE to ensure the existence of the engine apis.

My question was more "what is OPENSSL_NO_ENGINE actually for"? If it causes the engine apis to exist as stubs, then it's necessary but could be checked only in the implementation of the runtime check. If it's JUST for the existence of the engine APIs, I'm not sure we actually need it.

Reply:
> The feature probe doesn't need OPENSSL_NO_ENGINE to ensure the existence of the engine apis.

Yep makes sense.

> My question was more "what is OPENSSL_NO_ENGINE actually for"? If it causes the engine apis to exist as stubs, then it's necessary but could be checked only in the implementation of the runtime check. If it's JUST for the existence of the engine APIs, I'm not sure we actually need it.

Due to a lack of official documentation about OPENSSL_NO_ENGINE behavior I think we should code defensively and include it in the runtime check since its presence does mean that openssl was built with the `no-engine` flag.